src/algebra/algebra/unitization.lean

```diff
@@ -0,0 +1,235 @@
+`[non_unital_ring A] [module R A] [smul_comm_class R A A] [is_scalar_tower R A A]`) we construct
+the minimal unital `R`-algebra containing `A` as an ideal. This object `algebra.unitization R A` is
+a type synonym for `R × A` on which we place a different multiplicative structure, namely,
+`(r₁, a₁) * (r₂, a₂) = (r₁ * r₂, r₁ • a₂ + r₂ • a₁ + a₁ * a₂)` where the multiplicative identity
+is `(1, 0)`.
+
+Note, when `A` is a *unital* `R`-algebra, then `unitization R A` constructs a new multiplicative
+identity different from the old one, and so in general `unitization R A` and `A` will not be
+isomorphic even in the unital case. This approach actually has nice functorial properties.
+
+There is a natural coercion from `A` to `algebra.unitization R A` given by `λ a, (0, a)`, the image
+of which is a proper ideal (TODO), and when `R` is a field this ideal is maximal. Moreover,
+this ideal is always an essential ideal (it has nontrivial intersection with every other nontrivial
+ideal).
+
+Every non-unital algebra homomorphism from `A` into a *unital* `R`-algebra `B` has a unique
+extension to a (unital) algebra homomorphism from `algebra.unitization R A` to `B`.
+
+## Main definitions
+
+* `algebra.unitization R A`: the unitization of a non-unital `R`-algebra `A`.
+* `unitization.algebra`: the unitization of `A` as a (unital) `R`-algebra.
+* `unitization.coe_non_unital_alg_hom`: coercion as a non-unital algebra homomorphism.
+* `non_unital_alg_hom.to_alg_hom φ`: the extension of a non-unital algebra homomorphism `φ : A → B`
+  into a unital `R`-algebra `B` to an algebra homomorphism `algebra.unitization R A →ₐ[R] B`.
+
+## Main results
+
+* `non_unital_alg_hom.to_alg_hom_unique`: the extension is unique
+
+## TODO
+
+* prove the unitization operation is a functor between the appropriate categories
+* prove the image of the coercion is an essential ideal, maximal if scalars are a field.
+-/
+
+/-- The minimal unitization of a non-unital `R`-algebra `A`. This is just a type synonym for
+`R × A`.-/
+def algebra.unitization (R A : Type*) := R × A
+
+instance unitization.inhabited (R A : Type*) [inhabited R] [inhabited A] :
+  inhabited (algebra.unitization R A) :=
+{ default := (default, default) }
+
+/-- The explicit identity map between `R × A` and `algebra.unitization R A`. -/
+def to_unitization {R A : Type*} : R × A ≃ algebra.unitization R A := equiv.refl _
+
+variables {R A : Type*}
+
+open algebra
+
+namespace unitization
+
+instance [has_zero R] : has_coe_t A (unitization R A) := { coe := λ a, to_unitization (0, a) }
+
+@[simp]
+lemma coe_apply [has_zero R] (a : A) : (↑a : unitization R A) = to_unitization (0, a) := rfl
+
+lemma coe_injective [has_zero R] : function.injective (coe : A → unitization R A) :=
+λ a₁ a₂, by simp
+
+lemma decomp (x : unitization R A) : ∃ r a, x = to_unitization (r, a) :=
+⟨(to_unitization.symm x).1, (to_unitization.symm x).2,
+  by simp only [prod.mk.eta, equiv.apply_symm_apply]⟩
+
+section ring
```

Comment:
Can you generalize this to a `section semiring` for many of these results?

Reply:
yes, done. Although it appears we don't have a commutative version of `non_unital_semiring`, is that correct? It would be so nice if `comm` was a mix-in.